tutorials/pubsub-spring-dedup-messages/index.md

```diff
@@ -0,0 +1,263 @@
+
+In [application.properties](pubsub-spring/src/main/resources/application.properties):
+
+[embedmd]:# (pubsub-spring/src/main/resources/application.properties /.*Data coming/ /=subscriptionFromDataflow/)
+```properties
+# Data coming from Cloud Pub/Sub to a Spring Cloud Stream sink defined in the
+# Consumer bean `receiveDedupedMessagesFromDataflow`. The application
+# will create the Cloud Pub/Sub topic `topicFromDataflow` and subscription
+# `topicFromDataflow.subscriptionFromDataflow` if they do not exist.
+spring.cloud.stream.bindings.receiveDedupedMessagesFromDataflow-in-0.destination=topicFromDataflow
+spring.cloud.stream.bindings.receiveDedupedMessagesFromDataflow-in-0.group=subscriptionFromDataflow
+```
+
+### Construct a Pub/Sub message with custom attributes
+
+Cloud Pub/Sub lets you publish messages with custom attributes, which are key value pairs. To construct such a message in your Spring Boot application, you can specify the attributes as headers. 
+
+In [DataEntryController.java](pubsub-spring/src/main/java/com/google/example/DataEntryController.java):
+
+[embedmd]:# (pubsub-spring/src/main/java/com/google/example/DataEntryController.java java /  \/\/ Headers/ /;/)
+```java
+  // Headers become Pub/Sub message attributes.
+    Message<String> message = MessageBuilder
+      .withPayload(data)
+      .setHeader("key", key)
+      .build();
+```
+
+### Run the application
+To start your application, navigate to `pubsub-spring/` and run: 
+
+```shell script
+mvn spring-boot:run
+```
+
+Check that your app has started successfully by pointing your browser to `localhost:8080`. You should be able to send messages using the form there. At this point, Spring will have automatically created the Cloud Pub/Sub topic that you have specified in `application.properties` to publish your message to. You can view Publish Message Request Count and Publish Message Operation Count in the topic details in [Cloud Console for Pub/Sub Topic] to verify that publishing to Cloud Pub/Sub is successful.  
+
+## Start a Cloud Dataflow Job to Deduplicate Pub/Sub Messages
+
+As a middle process that takes data from a Cloud Pub/Sub topic and publishes processed data to another Cloud Pub/Sub topic, Cloud Dataflow achieves exactly once stream processing by asking the input stream for an `idAttribute`. The `idAttribute` can be a Pub/Sub message attribute. Cloud Dataflow will avoid processing messages that have the same values in this `idAttribute` repeatedly.
```

Comment:
It's not clear to me how being a middle process lets Dataflow achieves exactly-once processing. This paragraph suggests that it achieves exactly-once processing by getting the `idAttribute` from the input stream. 

We should also use "Dataflow" instead of "Cloud Dataflow" now.